### PR TITLE
Compatibility for `weights_only=True` by default

### PR DESCRIPTION
**Pull request by matsumotosan**
## What does this PR do?

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/pytorch/core/saving.py [hunk 1]
```diff
@@ -56,11 +56,17 @@ def _load_from_checkpoint(
     map_location: _MAP_LOCATION_TYPE = None,
     hparams_file: Optional[_PATH] = None,
     strict: Optional[bool] = None,
+    weights_only: Optional[bool] = None,
```
Comment (matsumotosan) [inline src/lightning/pytorch/core/saving.py:59]: Should we default to `weights_only=None` or `weights_only=True`? If we have no use for `weights_only=None`, we can simplify the type hint to `weights_only: bool = True`.

File: tests/tests_pytorch/checkpointing/test_legacy_checkpoints.py [hunk 2]
```diff
@@ -45,7 +46,12 @@ def test_load_legacy_checkpoints(tmp_path, pl_version: str):
         assert path_ckpts, f'No checkpoints found in folder "{PATH_LEGACY}"'
         path_ckpt = path_ckpts[-1]
 
-        model = ClassificationModel.load_from_checkpoint(path_ckpt, num_features=24)
+        # legacy load utility added in 1.5.0 (see https://github.com/Lightning-AI/pytorch-lightning/pull/9166)
+        if pl_version == "local":
```
Comment (matsumotosan) [inline tests/tests_pytorch/checkpointing/test_legacy_checkpoints.py:49]: This is the simplest way that I could think of ensuring we continue testing the legacy checkpoints. Another way could be to use [`torch.serialization.add_safe_globals`](https://docs.pytorch.org/docs/main/notes/serialization.html#torch.serialization.add_safe_globals), but it seems a little more complicated (particularly since we're using the `pl_legacy_patch` context manager already.

**Comment by matsumotosan** [comment]
@Borda I wanted to get your opinion on something before moving forward.

I've added `weights_only` as an argument to `LightningModule.load_from_checkpoint` and all downstream functions to allow users to determine which option they want to use to load checkpoints.

My issue right now is with resuming training from a checkpoint with `Trainer.fit`. I see a few options right now:
1. Add `weights_only` as an argument to `Trainer.fit` (would also have to modify args for `validate`, `test`, and `predict`). Set default value to `True`.
2. Use `weights_only=True` everywhere, and print an error message advising user to set `TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD` if they would like to load with `weights_only=False`. [Users must explicitly set environment variable to force loading with `weights_only=False`](https://github.com/pytorch/pytorch/blob/ba56102387ef21a3b04b357e5b183d48f0afefc7/torch/serialization.py#L1439).
3. Add `weights_only` as an argument to `Trainer` initialization. Easy, but would not allow fine-grained control on loading models between different calls of `fit`, `validate`, etc.

I'm leaning towards option 1, but it involves changing up `Trainer` methods, which affects a lot of code so wanted to run this by you beforehand.

**Comment by Borda** [comment]
> My issue right now is with resuming training from a checkpoint with `Trainer.fit`. I see a few options right now:
> 
> 1. Add `weights_only` as an argument to `Trainer.fit` (would also have to modify args for `validate`, `test`, and `predict`). Set default value to `True`.
> 2. Use `weights_only=True` everywhere, and print an error message advising user to set `TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD` if they would like to load with `weights_only=False`. [Users must explicitly set environment variable to force loading with `weights_only=False`](https://github.com/pytorch/pytorch/blob/ba56102387ef21a3b04b357e5b183d48f0afefc7/torch/serialization.py#L1439).
> 3. Add `weights_only` as an argument to `Trainer` initialization. Easy, but would not allow fine-grained control on loading models between different calls of `fit`, `validate`, etc.

The cleanest way would probably be 1), but it brings so many new arguments for a marginal use... so personally I would go with 2)
cc: @lantiga

**Comment by matsumotosan** [comment]
@Borda I will go ahead with changing all cases of `torch.load` to use `weights_only=True`.

This will cause a lot of errors with checkpoints from previous versions, so I'll update the docs/warning messages as well to inform users to use either the context manager or global environment variable.

**Comment by lantiga** [comment]
Hi @matsumotosan let's do that only if the underlying torch is >= 2.6 (since starting weights_only became True by default from that point on), otherwise we're going to break a lot of older code

**Comment by matsumotosan** [comment]
I am not sure if it's possible to default to `weights_only=True` for `trainer.{test,validate,test,predict}`, since it loads a checkpoint at some point and that checkpoint may include elements that are not allowed by `torch.load(..., weights_only=True)` (Lightning itself saves classes like `ModelCheckpoint`), which cannot be loaded unless `weights_only=False` or a context manager is used.

The big issue with context managers is that a different one has to be used each time a different checkpoint is loaded. Setting the environment variable `TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD=1` does not work as it is overriden by the value passed to `torch.load`.

With this in mind, I think passing `weights_only` to `trainer.{fit,validate,test,predict}` may be the only short-term solution.

If we need to force the `weights_only=True` solution, every item saved in the checkpoint would need to be converted into a primitive type that `torch.load(.., weights_only=True)` would accept.

I have also added `weights_only` as an argument to the `LightningModule` and `LightningDataModule` classes as there are a few issues that point this out as a source of error:
- https://github.com/Lightning-AI/pytorch-lightning/issues/20058
- https://github.com/Lightning-AI/pytorch-lightning/issues/20643

Maybe we could default add `weights_only` and have it default to `False` for now until a future release so that we don't break users' code whilst adding an explicit option for this?

File: src/lightning/fabric/utilities/imports.py [hunk 3]
```diff
@@ -35,6 +35,7 @@
 _TORCH_GREATER_EQUAL_2_4 = compare_version("torch", operator.ge, "2.4.0")
 _TORCH_GREATER_EQUAL_2_4_1 = compare_version("torch", operator.ge, "2.4.1")
 _TORCH_GREATER_EQUAL_2_5 = compare_version("torch", operator.ge, "2.5.0")
+_TORCH_GREATER_EQUAL_2_6 = compare_version("torch", operator.ge, "2.6.0")
```
Comment (Borda) [inline src/lightning/fabric/utilities/imports.py:38]: so, where is the use? I thought that we do the defaulting only for _TORCH_GREATER_EQUAL_2_6

File: src/lightning/fabric/utilities/cloud_io.py [hunk 4]
```diff
@@ -51,6 +56,9 @@ def _load(
             weights_only=weights_only,
         )
     if str(path_or_url).startswith("http"):
+        if weights_only is None:
```
Comment (Borda) [inline src/lightning/fabric/utilities/cloud_io.py:59]:
```suggestion
        if weights_only is None and _TORCH_GREATER_EQUAL_2_6:
```

Comment (matsumotosan) [reply src/lightning/fabric/utilities/cloud_io.py:59]: @Borda 

[`torch.load_state_dict_from_url`](https://docs.pytorch.org/docs/2.8/hub.html#torch.hub.load_state_dict_from_url) defaults to `weights_only=False` even for 2.8. Should we still default to `weights_only=True` in this case?

For cases of `torch.load`, I am not modifying `weights_only` if the function receives `weights_only=None` as the underlying `torch.load` function will handle defaults. We cannot do that with `torch.load_state_dict_from_url` as the argument is `weights_only: bool = False`.

Comment (Borda) [reply src/lightning/fabric/utilities/cloud_io.py:59]: ok, lest set our default only if there is none yet

File: src/lightning/fabric/CHANGELOG.md [hunk 5]
```diff
@@ -9,6 +9,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Added
 
+- Added `exclude_frozen_parameters` to `DeepSpeedStrategy` ([#21060](https://github.com/Lightning-AI/pytorch-lightning/pull/21060))
```
Comment (Borda) [inline src/lightning/fabric/CHANGELOG.md]:
```suggestion
```

File: src/lightning/fabric/CHANGELOG.md [hunk 6]
```diff
@@ -19,7 +20,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Changed
 
--
+- Default to `weights_only=True` for `torch>=2.6` when loading checkpoints. ([#21072](https://github.com/Lightning-AI/pytorch-lightning/pull/21072))
+- let `_get_default_process_group_backend_for_device` support more hardware platforms ([#21057](https://github.com/Lightning-AI/pytorch-lightning/pull/21057), [#21093](https://github.com/Lightning-AI/pytorch-lightning/pull/21093))
```
Comment (Borda) [inline src/lightning/fabric/CHANGELOG.md]:
```suggestion
```

File: pyproject.toml [hunk 7]
```diff
@@ -182,6 +182,8 @@ filterwarnings = [
     # "error::DeprecationWarning",
     "error::FutureWarning",
     "ignore::FutureWarning:onnxscript", # Temporary ignore until onnxscript is updated
+    "ignore:You are using `torch.load` with `weights_only=False`.*:FutureWarning",
+    "ignore:The pynvml package is deprecated:FutureWarning", # Ignore pynvml deprecation warning, since it is not installed by PL directly
```
Comment (Borda) [inline pyproject.toml]:
```suggestion
```

File: src/lightning/fabric/strategies/model_parallel.py [hunk 8]
```diff
@@ -411,6 +411,7 @@ def _load_checkpoint(
     state: dict[str, Union[Module, Optimizer, Any]],
     strict: bool = True,
     optimizer_states_from_list: bool = False,
+    weights_only: bool = False,
```
Comment (Borda) [inline src/lightning/fabric/strategies/model_parallel.py]: I thought we would keep it unset as None until the very lat point we need to pas a bool value so we wont add any assumption to the user code

File: src/lightning/pytorch/core/saving.py (hunk 1 above)
Comment (Borda) [reply src/lightning/pytorch/core/saving.py:59]: I think for compatibility default to None

File: src/lightning/fabric/strategies/model_parallel.py (hunk 8 above)
Comment (matsumotosan) [reply src/lightning/fabric/strategies/model_parallel.py]: Yes, that is the idea. Thank you for catching this.

File: src/lightning/pytorch/trainer/connectors/checkpoint_connector.py [hunk 9]
```diff
@@ -403,18 +403,22 @@ def restore_lr_schedulers(self) -> None:
         for config, lrs_state in zip(self.trainer.lr_scheduler_configs, lr_schedulers):
             config.scheduler.load_state_dict(lrs_state)
 
-    def _restore_modules_and_callbacks(self, checkpoint_path: Optional[_PATH] = None) -> None:
+    def _restore_modules_and_callbacks(
+        self, checkpoint_path: Optional[_PATH] = None, weights_only: Optional[bool] = None
+    ) -> None:
         # restore modules after setup
-        self.resume_start(checkpoint_path)
+        self.resume_start(checkpoint_path, weights_only)
```
Comment (Borda) [inline src/lightning/pytorch/trainer/connectors/checkpoint_connector.py]:
```suggestion
        self.resume_start(checkpoint_path, weights_only=weights_only)
```

File: src/lightning/pytorch/trainer/connectors/checkpoint_connector.py [hunk 10]
```diff
@@ -80,7 +80,7 @@ def resume_start(self, checkpoint_path: Optional[_PATH] = None) -> None:
 
         rank_zero_info(f"Restoring states from the checkpoint path at {checkpoint_path}")
         with pl_legacy_patch():
-            loaded_checkpoint = self.trainer.strategy.load_checkpoint(checkpoint_path)
+            loaded_checkpoint = self.trainer.strategy.load_checkpoint(checkpoint_path, weights_only)
```
Comment (Borda) [inline src/lightning/pytorch/trainer/connectors/checkpoint_connector.py]:
```suggestion
            loaded_checkpoint = self.trainer.strategy.load_checkpoint(checkpoint_path, weights_only=weights_only)
```

File: src/lightning/pytorch/trainer/trainer.py [hunk 11]
```diff
@@ -676,14 +686,15 @@ def validate(
...
-            self, self._validate_impl, model, dataloaders, ckpt_path, verbose, datamodule
+            self, self._validate_impl, model, dataloaders, ckpt_path, weights_only, verbose, datamodule
         )
 
     def _validate_impl(
         self,
         model: Optional["pl.LightningModule"] = None,
         dataloaders: Optional[Union[EVAL_DATALOADERS, LightningDataModule]] = None,
         ckpt_path: Optional[_PATH] = None,
+        weights_only: Optional[bool] = None,
```
Comment (Borda) [inline src/lightning/pytorch/trainer/trainer.py]: lets move it to the last to prevent breaking chnages

File: src/lightning/pytorch/trainer/trainer.py [hunk 12]
```diff
@@ -621,6 +630,7 @@ def validate(
...
```
Comment (Borda) [inline src/lightning/pytorch/trainer/trainer.py]: lets move it to the last to prevent breaking changes

File: src/lightning/pytorch/trainer/trainer.py [hunk 13]
```diff
@@ -899,7 +913,7 @@ def predict(
         self.state.status = TrainerStatus.RUNNING
         self.predicting = True
         return call._call_and_handle_interrupt(
-            self, self._predict_impl, model, dataloaders, datamodule, return_predictions, ckpt_path
+            self, self._predict_impl, model, dataloaders, datamodule, return_predictions, ckpt_path, weights_only
```
Comment (Borda) [inline src/lightning/pytorch/trainer/trainer.py]:
```suggestion
            self, self._predict_impl, model, dataloaders, datamodule, return_predictions, ckpt_path, weights_only=weights_only
```

**Comment by deependujha** [comment]
Seems like an actual issue than a flaky test behavior
